numba/targets/arrayobj.py

```diff
@@ -3756,70 +3756,95 @@ def take_impl(a, indices):
     return impl_ret_new_ref(context, builder, sig.return_type, res)
 
 
-@lower_builtin(np.arange, types.Number)
-def numpy_arange_1(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
-
-    def arange(stop):
-        return np.arange(0, stop, 1, dtype)
-
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
-
+def _arange_dtype(*args):
+    bounds = [a for a in args if not isinstance(a, types.NoneType)]
 
-@lower_builtin(np.arange, types.Number, types.Number)
-def numpy_arange_2(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
+    if any(isinstance(a, types.Complex) for a in bounds):
+        dtype = types.complex128
+    elif any(isinstance(a, types.Float) for a in bounds):
+        dtype = types.float64
+    else:
+        dtype = max(bounds)
 
-    def arange(start, stop):
-        return np.arange(start, stop, 1, dtype)
+    return dtype
 
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
 
+@register_jitable
+def _arange_impl_complex(_start, _stop, _step, dtype):
+    step = _step if _step is not None else 1
+    if _stop is None:
+        start, stop = 0, _start
+    else:
+        start, stop = _start, _stop
 
-@lower_builtin(np.arange, types.Number, types.Number,
-               types.Number)
-def numpy_arange_3(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
-
-    def arange(start, stop, step):
-        return np.arange(start, stop, step, dtype)
+    nitems_c = (stop - start) / step
+    nitems_r = int(math.ceil(nitems_c.real))
+    nitems_i = int(math.ceil(nitems_c.imag))
+    nitems = max(min(nitems_i, nitems_r), 0)
+    arr = np.empty(nitems, dtype)
+    val = start
+    for i in range(nitems):
+        arr[i] = val
+        val += step
+    return arr
 
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
 
+@register_jitable
+def _arange_impl_real(_start, _stop, _step, dtype):
+    step = _step if _step is not None else 1
+    if _stop is None:
+        start, stop = 0, _start
+    else:
+        start, stop = _start, _stop
+
+    nitems_r = int(math.ceil((stop - start) / step))
+    nitems = max(nitems_r, 0)
+    arr = np.empty(nitems, dtype)
+    val = start
+    for i in range(nitems):
+        arr[i] = val
+        val += step
+    return arr
+
+
+@overload(np.arange)
+def np_arange(start, stop=None, step=None, dtype=None):
```

Comment:
Should the default value for `step` be `None` or `1`? IIRC it's 1.

Reply:
I can't find the declaration of `np.arange` in the numpy source code, but the documentation suggets it is 1.  However, this works, so I feel like it may be `None`.
```
>>> np.arange(10, step=None)
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```
This use case is exercised by the unit tests I've written.

Reply:
https://github.com/numpy/numpy/blob/c592bf2210567ea58bc52f8e15a5f0de582f889b/numpy/core/src/multiarray/ctors.c#L3493-L3498, if `None` sets step to 1. Think your patch does the same so this seems ok?